integration/test/Testlib/App.hs

```diff
@@ -0,0 +1,935 @@
+  (ProducesJSON a, ProducesJSON b, HasCallStack) =>
+  -- | The actual value
+  a ->
+  -- | The expected value
+  b ->
+  App ()
+a `shouldMatch` b = do
+  xa <- prodJSON a
+  xb <- prodJSON b
+  unless (xa == xb) $ do
+    pa <- prettyJSON xa
+    pb <- prettyJSON xb
+    assertFailure $ "Expected:\n" <> pb <> "\n" <> "Actual:\n" <> pa
+
+shouldMatchInt ::
+  (ProducesJSON a, HasCallStack) =>
+  -- | The actual value
+  a ->
+  -- | The expected value
+  Int ->
+  App ()
+a `shouldMatchInt` b = do
+  xa <- prodJSON a
+  shouldMatch xa b
+
+liftP2 ::
+  (ProducesJSON a, ProducesJSON b, HasCallStack) =>
+  (Value -> Value -> c) ->
+  a ->
+  b ->
+  App c
+liftP2 f a b = do
+  f <$> prodJSON a <*> prodJSON b
+
+isEqual ::
+  (ProducesJSON a, ProducesJSON b, HasCallStack) =>
+  a ->
+  b ->
+  App Bool
+isEqual = liftP2 (==)
+
+printFailureDetails :: AssertionFailure -> ResultDetailsPrinter
+printFailureDetails (AssertionFailure stack mbResponse _) = ResultDetailsPrinter $ \testLevel _withFormat -> do
+  let nindent = 2 * testLevel + 2
+  putStrLn (indent nindent (prettyStack stack))
+  for_ mbResponse $ \r -> putStrLn (indent nindent (prettyReponse r))
+
+prettyStack :: CallStack -> String
+prettyStack cstack =
+  intercalate "\n" $
+    [colored yellow "call stack: "]
+      <> (drop 1 . prettyCallStackLines) cstack
+
+modifyFailure :: (AssertionFailure -> AssertionFailure) -> App a -> App a
+modifyFailure modifyAssertion action = do
+  env <- ask
+  liftIO
+    ( E.catch
+        (runAppWithEnv env action)
+        ( \(e :: AssertionFailure) ->
+            E.throw (modifyAssertion e)
+        )
+    )
+
+modifyFailureMsg :: (String -> String) -> App a -> App a
+modifyFailureMsg modMessage = modifyFailure (\e -> e {msg = modMessage e.msg})
+
+addFailureContext :: String -> App a -> App a
+addFailureContext msg = modifyFailureMsg (\m -> m <> "\nThis failure happend in this context:\n" <> msg)
+
+-------------------------------------------------------------------------------
+-- - SECTION_JSON
+-------------------------------------------------------------------------------
+
+-- | All library functions should use this typeclass for all untyped value
+-- arguments wherever possible. This design choice has advantages:
+--
+-- No need convert value between different representations. E.g. if a function
+-- needs a user id as a string, all these input types become valid input:
+--
+-- - String
+-- - Text
+-- - Value
+-- - App Text
+-- - App String
+-- - App Value
+--
+-- Internally the function calls `asString` to convert to App String
+--
+-- Since (App a) are treated as first-class values values this means we can
+-- compose operations that might fail without giving up nice error messages:
+--
+-- callMe (response.json %. "user" & "foo.bar.baz" %.= 2)
+--
+-- This can fail if
+-- 1. the response is not application/json
+-- 2. has no "user" field
+-- 3. the nested update fails
+class ProducesJSON a where
+  prodJSON :: HasCallStack => a -> App Value
```

Comment:
```suggestion
class MakeSON a where
  makeJSON :: HasCallStack => a -> App Value
```

or `mk`?  or `gen`?

Would be nice if method and class name would be consistent, and `produce` is a bit long for something so heavily used.

Reply:
Renamed it to `MakesValue` and `make`